core/src/storage/collections/binary_heap/impls.rs

```diff
@@ -0,0 +1,448 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+// This file is part of ink!.
+//
+// ink! is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// ink! is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with ink!.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::duplex_sync_chunk::DuplexSyncChunk;
+use crate::storage::{
+    self,
+    alloc::{
+        Allocate,
+        AllocateUsing,
+        Initialize,
+    },
+    chunk::SyncChunk,
+    Flush,
+};
+use core::cmp::{
+    Ord,
+    Ordering,
+};
+#[cfg(feature = "ink-generate-abi")]
+use ink_abi::{
+    HasLayout,
+    LayoutField,
+    LayoutStruct,
+    StorageLayout,
+};
+use scale::{
+    Codec,
+    Decode,
+    Encode,
+};
+#[cfg(feature = "ink-generate-abi")]
+use type_metadata::Metadata;
+
+/// We implement a binary tree.
+pub const CHILDREN: u32 = 2;
+
+/// A binary heap collection.
+/// The heap depends on `Ord` and is a max-heap by default. In order to
+/// make it a min-heap implement the `Ord` trait explicitly on the type
+/// which is stored in the heap.
+///
+/// Provides `O(log(n))` push and pop operations.
+#[derive(Debug)]
+#[cfg_attr(feature = "ink-generate-abi", derive(Metadata))]
+pub struct BinaryHeap<T> {
+    /// The number of nodes stored in the heap.
+    len: storage::Value<u32>,
+    /// The nodes of the heap.
+    entries: DuplexSyncChunk<T>,
```

Comment:
Yeah, so the layout is currently not ideal. This was just the lowest hanging fruit ‒ taking the "array" and splitting it into groups to save together without further optimization of grouping.

For [this binary max heap](https://en.wikipedia.org/wiki/File:Max-Heap.svg) the layout is currently as follows (three nodes in each group):
```
Group 0: [100, 19, 36]
Group 1: [17, 3, 25]
Group 2: [1, 2, 7]
```
(I can also add a test and comment for this once we know which direction we're going.)

Hmm, maybe it would be an idea to always keep the root in it's own group 0 and then batch pairs of two nodes together. But really, it's all just premature optimization. Do you have any thoughts?

(Btw, the `BTreeMap` algorithm will do this optimization for reads a lot better, since that data structure inherently is suited for minimizing reads.)